### PR TITLE
Seperate client and server of FTP

### PR DESCRIPTION
**Pull request by AlexDvorak**
fixes issues found in PR#1044 @cclauss

### CODE REVIEW

**Comment by AlexDvorak** [comment]
TravisCI is hanging on pytest probably due to the server running an infinite loop, should we add it to the ignore in the pytest command, or should I make it so it only accepts one connection? @cclauss 
```
============================= test session starts ==============================

platform linux -- Python 3.7.1, pytest-4.3.1, py-1.7.0, pluggy-0.8.0

rootdir: /home/travis/build/TheAlgorithms/Python, inifile:

collecting 22 items                                                            

No output has been received in the last 10m0s, this potentially indicates a stalled build or something wrong with the build itself.

Check the details on how to adjust your build configuration on: https://docs.travis-ci.com/user/common-build-problems/#build-times-out-because-no-output-was-received

The build has been terminated
```

File: file_transfer_protocol/mytext.txt [hunk 1]
```diff
@@ -0,0 +1,2 @@
+Hello
+This is sample data
```
Comment (cclauss) [inline file_transfer_protocol/mytext.txt]: Let’s add Unicode data like Lüsai and 😄

File: file_transfer_protocol/client.py [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+    while True:
+        print('receiving data...')
+        data = s.recv(1024)
+        print('data=%s', (data))
+        if not data:
+            break
+        # write data to a file
+        f.write(data)
+
+f.close()
```
Comment (cclauss) [inline file_transfer_protocol/client.py]: __f__ is created by a __with__ context manager which automatically calls _f.close()__ so doing that again is not required.

**Comment by cclauss** [comment]
Also remove this line and make sure the Travis test pass.
https://github.com/TheAlgorithms/Python/blob/master/.travis.yml#L15

File: file_transfer_protocol/mytext.txt (hunk 1 above)
Comment (AlexDvorak) [reply file_transfer_protocol/mytext.txt]: Sure, sounds good

File: file_transfer_protocol/client.py (hunk 2 above)
Comment (AlexDvorak) [reply file_transfer_protocol/client.py]: alright, will fix

File: file_transfer_protocol/client.py [hunk 3]
```diff
@@ -7,18 +7,17 @@
...
-with open('received_file', 'wb') as f:
-    print('file opened')
+with open('Received_file', 'wb') as f:
+    print('File opened')
+    print('Receiving data...')
     while True:
-        print('receiving data...')
         data = s.recv(1024)
-        print('data=%s', (data))
+        print('data='+str(data))
```
Comment (cclauss) [inline file_transfer_protocol/client.py]: Let's use an __f"string"__ instead.

File: .travis.yml [hunk 4]
```diff
@@ -12,8 +12,6 @@ script:
   - pytest . --doctest-modules
       --ignore=data_structures/stacks/balanced_parentheses.py
       --ignore=data_structures/stacks/infix_to_postfix_conversion.py
-      --ignore=file_transfer_protocol/ftp_send_receive.py
```
Comment (cclauss) [inline .travis.yml]: Was __ftp_send_receive.py__ fixed?

Comment (AlexDvorak) [reply .travis.yml]: I did not make any changes, but will test later

File: file_transfer_protocol/server.py [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+while True:
+    conn, addr = s.accept()     # Establish connection with client.
+    print('Got connection from', addr)
+    data = conn.recv(1024)
+    print('Server received', repr(data))
+
+    filename='mytext.txt'
+    f = open(filename,'rb')
+    l = f.read(1024)
+    while (l):
```
Comment (cclauss) [inline file_transfer_protocol/server.py]: L is a VERY ambiguous identifier.  Is it a one or an ell.  Please avoid single character variable names like the plague that they are.

File: file_transfer_protocol/client.py [hunk 6]
```diff
@@ -1,13 +1,13 @@
...
 host = socket.gethostname()     # Get local machine name
 port = 12312
 
-s.connect((host, port))
-s.send(b'Hello server!')
+sock.connect((host, port))
+sock.send(b'Hello server!')
 
-with open('Received_file', 'wb') as f:
+with open('Received_file', 'wb') as file:
```
Comment (cclauss) [inline file_transfer_protocol/client.py]: Please change __file__ to __out_file__ to make the filemode clear to the reader.

File: file_transfer_protocol/server.py [hunk 7]
```diff
@@ -3,32 +3,32 @@
...
 
     filename='mytext.txt'
-    f = open(filename,'rb')
-    l = f.read(1024)
-    while (l):
-       conn.send(l)
-       print('Sent ',repr(l))
-       l = f.read(1024)
-    f.close()
+    file = open(filename,'rb')
```
Comment (cclauss) [inline file_transfer_protocol/server.py]: Please change __file__ to __in_file__ to make the filemode clear to the reader.

Comment (cclauss) [reply file_transfer_protocol/server.py]: Use the __with open() as__ syntax to ensure that the file.close() happens even if an exception is thrown.

**Comment by AlexDvorak** [comment]
(this is a seperate issue, but the actual FTP protocol does not operate as it does in `client.py` and `server.py`)

**Comment by cclauss** [comment]
One __undefined name 's'__ https://travis-ci.org/TheAlgorithms/Python/jobs/568640601#L564

**Comment by cclauss** [comment]
It seems to be hanging our pytest discovery process.  Perhaps put the code under an ìf __name__ == "__main__":` block so that is not executed each time that pytest imports the file.

**Comment by AlexDvorak** [comment]
the entirity of `client.py` and `server.py`?

File: file_transfer_protocol/server.py [hunk 8]
```diff
@@ -0,0 +1,34 @@
+import socket                   # Import socket module
+
+ONE_CONNECTION_ONLY = True      # Set this to False if you wish to continuously accept connections
+
+filename='mytext.txt'
+port = 12312                    # Reserve a port for your service.
+sock = socket.socket()             # Create a socket object
+host = socket.gethostname()     # Get local machine name
+sock.bind((host, port))            # Bind to the port
+sock.listen(5)                     # Now wait for client connection.
```
Comment (cclauss) [inline file_transfer_protocol/server.py]: Either line up these comments or consistently use `<space><space>#<space>`

File: file_transfer_protocol/client.py [hunk 9]
```diff
@@ -0,0 +1,23 @@
+import socket                   # Import socket module
+
+sock = socket.socket()             # Create a socket object
+host = socket.gethostname()     # Get local machine name
+port = 12312
+
+sock.connect((host, port))
+sock.send(b'Hello server!')
+
+with open('Received_file', 'wb') as in_file:
```
Comment (cclauss) [inline file_transfer_protocol/client.py]: We write data to an out_file.

File: file_transfer_protocol/server.py [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+
+print('Server listening....')
+
+while True:
+    conn, addr = sock.accept()     # Establish connection with client.
+    print('Got connection from', addr)
+    data = conn.recv(1024)
+    print('Server received', repr(data))
+
+    with open(filename,'rb') as out_file:
```
Comment (cclauss) [inline file_transfer_protocol/server.py]: We read data from an in_file.

File: file_transfer_protocol/server.py [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+        data = out_file.read(1024)
+        while (data):
+           conn.send(data)
+           print('Sent ',repr(data))
```
Comment (cclauss) [inline file_transfer_protocol/server.py]: f"string"

File: file_transfer_protocol/client.py (hunk 9 above)
Comment (AlexDvorak) [reply file_transfer_protocol/client.py]: sorry, I was confused, I was thinking it was sending it out (like a physical thing)

File: file_transfer_protocol/server.py [hunk 12]
```diff
@@ -0,0 +1,34 @@
...
+           data = out_file.read(1024)
+
+    print('Done sending')
+    conn.send(b'Thank you for connecting')
```
Comment (cclauss) [inline file_transfer_protocol/server.py]: Is this data going to be put into the output file on the other end?

File: file_transfer_protocol/server.py (hunk 11 above)
Comment (cclauss) [reply file_transfer_protocol/server.py]: print('Sent {dataf!r}')

File: file_transfer_protocol/server.py (hunk 12 above)
Comment (AlexDvorak) [reply file_transfer_protocol/server.py]: no, it should not, I will fix that

**Comment by AlexDvorak** [comment]
what should we do about this https://travis-ci.org/TheAlgorithms/Python/jobs/568644960#L312
`ERROR: file not found: pytest`

**Comment by cclauss** [comment]
Landed!!  Well done!

**Pull request merged**